src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ConnectionPool/TransactedConnectionPool.cs

```diff
@@ -0,0 +1,348 @@
+                    }
+                }
+            }
+            finally
+            {
+                if (transactionClone != null)
+                {
+                    if (newConnections != null)
+                    {
+                        // another thread created the transaction pool and thus the new 
+                        //   TransactedConnectionList was not used, so dispose of it and
+                        //   the transaction clone that it incorporates.
+                        newConnections.Dispose();
+                    }
+                    else
+                    {
+                        // memory allocation for newConnections failed...clean up unused transactionClone
+                        transactionClone.Dispose();
+                    }
+                }
+            }
+            SqlClientEventSource.Log.TryPoolerTraceEvent("<prov.DbConnectionPool.TransactedConnectionPool.PutTransactedObject|RES|CPOOL> {0}, Transaction {1}, Connection {2}, Added.", Id, transaction.GetHashCode(), transactedObject.ObjectID);
+        }
+
+        SqlClientEventSource.Metrics.EnterFreeConnection();
+    }
+
+    /// <summary>
+    /// Handles the completion of a transaction by removing the associated connection from the 
+    /// transacted pool and returning it to the main connection pool.
+    /// </summary>
+    /// <param name="transaction">The transaction that has completed.</param>
+    /// <param name="transactedObject">The database connection that was enlisted in the completed transaction.</param>
+    /// <remarks>
+    /// This method is called when a transaction completes (either by committing or rolling back).
+    /// It removes the specified connection from the transacted pool and returns it to the main
+    /// connection pool for reuse. If this was the last connection in the transaction's pool,
+    /// the entire transaction pool entry is removed and disposed.
+    /// 
+    /// Due to the asynchronous nature of transaction completion notifications, there's no guarantee
+    /// about the order in which PutTransactedObject and TransactionEnded are called. This method
+    /// handles cases where the transaction pool may not yet exist when the transaction completes.
+    /// </remarks>
+    internal void TransactionEnded(Transaction transaction, DbConnectionInternal transactedObject)
+    {
+        SqlClientEventSource.Log.TryPoolerTraceEvent("<prov.DbConnectionPool.TransactedConnectionPool.TransactionEnded|RES|CPOOL> {0}, Transaction {1}, Connection {2}, Transaction Completed", Id, transaction.GetHashCode(), transactedObject.ObjectID);
+        TransactedConnectionList? connections;
+        int entry = -1;
+
+        // NOTE: because TransactionEnded is an asynchronous notification, there's no guarantee
+        //   around the order in which PutTransactionObject and TransactionEnded are called. As
+        //   such, it is possible that the transaction does not yet have a pool created.
```

Comment:
Maybe for later: it might be worth clarifying whether this is still accurate, and which pool it's referring to. This is only ever called if the connection's pool and the pool's `_transactedConnectionPool` variable is non-null.